# VIral GEnome ASsembly pipelines for WGS

This repository contains scripts and files to run the bioinformatic analysis of whole genome sequencing of viruses using Illumina or Oxford Nanopore Technologies platforms.

## INSTALLATION

### Linux/Unix

```sh
git clone --recursive https://github.com/khourious/vigeas.git && cd vigeas
chmod 700 -R INSTALL_Unix scripts
bash INSTALL_Unix
```

## INPUT DATA FORMATS

### Oxford Nanopore Technologies

POD5 is the current raw data format generated by ONT sequencing devices. For basecalling and demultiplexing, we provide an [IPython notebook](https://github.com/khourious/vigeas/tree/main/google_colab) that utilizes GPU acceleration for faster processing, which can be used as an alternative if you don't have access to local GPU resources.

For **older raw FAST5 data**, ONT provides a tool to convert FAST5 files to POD5: [https://pod5.nanoporetech.com/](https://pod5.nanoporetech.com)

This workflow expects demultiplexed data in BAM format. Ensure your input folder contains BAM files named with their respective barcodes (_e.g._, `barcode01.bam`, `barcode02.bam`). If your data is in FASTQ format, use the following conversion:
```sh
samtools import -0 input.fastq.gz -o output.bam
```

Prepare a sample sheet in CSV format with the following columns:
- **sample_id**: sample identifier
- **barcode**: barcode ID in BC format (_e.g._, `BC01`, `BC02`)
- **primer_scheme**: primer scheme name including version (_e.g._, `ChikAsianECSA_V1`)

Example:
```sh
976530,BC01,ARTIC_V4
985322,BC02,ZikaAsian_V2
```

Available primer schemes in this workflow can be found at: [primer_schemes/README.md](https://github.com/khourious/vigeas/blob/main/primer_schemes/README.md)

## USAGE

```
Usage: vigeas <command> or <miscellaneous>

Commands:
  ill   For Illumina Sequencing [*.fastQ data]
  ont   For ONT Sequencing [*.pod5 data]

Miscellaneous:
  clr3     List supported Clair3 models
  makedb   Create a BLAST database in this workflow -- for <vigeas ill -x hyb>
  panels   List available enrichment panels in this workflow -- for <vigeas ill -x hyb>
  primers  List available primer schemes in this workflow -- for <vigeas ill -x amp> and <vigeas ont -x bda>
  update   Update conda/mamba dependencies
  version  Show last update information
```

## CITATION

## CONTRIBUTIONS
Thanks to:
- Ricardo Khouri ([Rkhour0](https://github.com/orgs/khourious/people/Rkhour0)) for mentorship and discussions on improving in the `-x amp` and `-x ill` workflows.
- Verity Hill ([ViralVerity](https://github.com/ViralVerity)) for her collaborative troubleshooting regarding low-coverage sequencing issues, primer trimming, and bioinformatic pipeline optimizations for DENV, ZIKV, and CHIKV data.
